pkg/server/registry/images/tag.go

```diff
@@ -69,26 +86,48 @@ func (t *TagStrategy) ImageTag(ctx context.Context, namespace, imageName string,
 	}
 
 	for _, img := range imageList.Items {
-		if img.Name == image.Name {
-			continue
-		}
 		for i, tag := range img.Tags {
 			imageParsedTag, err := name.NewTag(tag, name.WithDefaultRegistry(""))
 			if err == nil {
 				tag = imageParsedTag.Name()
 			} else {
 				logrus.Errorf("invalid tag [%s] found, ignoring: %v", tag, err)
 			}
-			if duplicateTag[tag] {
+			if duplicateTag[tag] && img.Name != image.Name {
 				img.Tags = append(img.Tags[:i], img.Tags[i+1:]...)
-				err := t.client.Update(ctx, &img)
+				result = append(result, img)
 				if err != nil {
-					return image, err
+					return nil, err
 				}
 			}
+			if duplicateTag[tag] && img.Name == image.Name {
+				removeDuplicateTag(tag, tags)
+			}
 		}
 	}
+	return result, nil
+}
 
-	image.Tags = append(image.Tags, tags...)
-	return image, t.client.Update(ctx, image)
+func removeDuplicateTag(tagToFind string, tags *[]string) {
+	for i, tag := range *tags {
+		imageParsedTag, err := name.NewTag(tag, name.WithDefaultRegistry(""))
+		if err != nil {
+			continue
+		}
+		if imageParsedTag.Name() == tagToFind {
+			*tags = append((*tags)[:i], (*tags)[i+1:]...)
+		}
+	}
+}
+
+func normalizeTags(tags []string) []string {
+	result := []string{}
```

Comment:
Small nit, but it's good practice to not allocate empty slices as they just immediately get thrown away once you append to them.  So it's just wasted memory allocation.
```suggestion
	var result []string
```